### PR TITLE
Union OCR / PDFMiner Tokens with Table Outputs

### PR DESCRIPTION
**Pull request by karanataryn**
We currently drop all table outputs that are not matched to tokens, but this is not required since we anyways chop the image after the DETR model. With this we aim to union back in outputs that did not fit into the original table structure, either by matching to existing cells, or creating new rows and columns for them. This is hidden behind a feature flag `table_extractor_options` that by default does not enable this until we are sure of no regressions.

### CODE REVIEW

**Comment by HenryL27** [comment]
```python
import sycamore
from sycamore import ExecMode
from sycamore.transforms.partition import ArynPartitioner
from sycamore.utils.pdf_utils import show_pages
import pandas as pd

file = "/Users/hmlin/Aryn/sycamore/debugging/CT-19919_NEXVIADYME_PIC_INS_AvisDef_CT19919_en.pdf"

ctx = sycamore.init(exec_mode=ExecMode.LOCAL)

ds = ctx.read.binary(paths=file, binary_format="pdf").partition(
    ArynPartitioner(
        use_ocr=True,
        extract_table_structure=True,
        use_partitioning_service=False,
    )
)
docs = ds.take_all()
elts = docs[0].elements
elts = [e for e in elts if e.type == "table"]
pd.set_option("max_colwidth", None)
for e in elts[:1]:
    print("=" * 80)
    print(e.table.to_pandas())
```

->

```
  File "/Users/hmlin/Aryn/sycamore/lib/sycamore/sycamore/transforms/table_structure/table_transformers.py", line 1104, in <genexpr>
    BoundingBox(*columns[column_num]["bbox"]) for column_num in token_columns
                 ~~~~~~~^^^^^^^^^^^^
TypeError: list indices must be integers or slices, not dict
```

**Comment by karanataryn** [comment]
> ```python
> import sycamore
> from sycamore import ExecMode
> from sycamore.transforms.partition import ArynPartitioner
> from sycamore.utils.pdf_utils import show_pages
> import pandas as pd
> 
> file = "/Users/hmlin/Aryn/sycamore/debugging/CT-19919_NEXVIADYME_PIC_INS_AvisDef_CT19919_en.pdf"
> 
> ctx = sycamore.init(exec_mode=ExecMode.LOCAL)
> 
> ds = ctx.read.binary(paths=file, binary_format="pdf").partition(
>     ArynPartitioner(
>         use_ocr=True,
>         extract_table_structure=True,
>         use_partitioning_service=False,
>     )
> )
> docs = ds.take_all()
> elts = docs[0].elements
> elts = [e for e in elts if e.type == "table"]
> pd.set_option("max_colwidth", None)
> for e in elts[:1]:
>     print("=" * 80)
>     print(e.table.to_pandas())
> ```
> 
> ->
> 
> ```
>   File "/Users/hmlin/Aryn/sycamore/lib/sycamore/sycamore/transforms/table_structure/table_transformers.py", line 1104, in <genexpr>
>     BoundingBox(*columns[column_num]["bbox"]) for column_num in token_columns
>                  ~~~~~~~^^^^^^^^^^^^
> TypeError: list indices must be integers or slices, not dict
> ```

Should be fixed in the most recent update.

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 1]
```diff
@@ -1092,11 +1094,11 @@ def structure_to_cells(table_structure, tokens, union_tokens=True):
                             token_columns.append(col_idx + 1)
                             break
                 if not token_rows:
-                    token_rows.append(max(rows) + 1)
+                    token_rows.append(len(rows) + 1)
```
[the remark this answers is not in the document]
Comment (HenryL27) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1097]: I got an out of range error I think that could do it

[the remark this answers is not in the document]
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1097]: Good catch, changed.

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 2]
```diff
@@ -1045,7 +1045,74 @@ def structure_to_cells(table_structure, tokens):
...
+                    if row_rect.intersect(token_rect).area > 0:
+                        token_rows.append(row)
+                    elif row_rect.y2 < token_rect.y1:
+                        if row_idx < len(rows) - 1 and rows[row_idx + 1]["bbox"][1] > token_rect.y2:
+                            new_row = BoundingBox(row_rect.x1, row_rect.y2, row_rect.x2, rows[row_idx + 1]["bbox"][1])
+                            rows.insert(row_idx + 1, {"bbox": new_row.to_list()})
+                            for cell in cells:
+                                for idx, row_num in enumerate(cell["row_nums"]):
+                                    if row_num > row_idx:
+                                        cell["row_nums"][idx] += 1
```
[the remark this answers is not in the document]
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1077]: For the `>=`, I don't think so? We add the new index at `row_idx + 1`. Will add the span check.

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 3]
```diff
@@ -1045,7 +1045,74 @@ def structure_to_cells(table_structure, tokens):
...
+                    elif col_rect.x2 < token_rect.x1:
+                        if col_idx < len(columns) - 1 and columns[col_idx + 1]["bbox"][0] > token_rect.x2:
+                            new_col = BoundingBox(
+                                col_rect.x2, col_rect.y1, columns[col_idx + 1]["bbox"][0], col_rect.y2
+                            )
+                            columns.insert(col_idx + 1, {"bbox": new_col.to_list()})
+                            for cell in cells:
+                                for idx, col_num in enumerate(cell["column_nums"]):
+                                    if col_num > col_idx:
+                                        cell["column_nums"][idx] += 1
```
[the remark this answers is not in the document]
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py]: Done.

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 4]
```diff
@@ -1045,7 +1045,74 @@ def structure_to_cells(table_structure, tokens):
         cell_rect = row_rect.intersect(column_rect)
         if cell_rect.area > 0:
             cell["bbox"] = cell_rect.to_list()
-
+    if union_tokens:
+        dropped_tokens = [
+            token for token, package_assignment in zip(tokens, package_assignments) if not package_assignment
+        ]
```
[the remark this answers is not in the document]
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1053]: Don't think it's needed? We should adjust for the successive elements regardless of the order of the list.

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 5]
```diff
@@ -1072,9 +1072,19 @@ def structure_to_cells(table_structure, tokens, union_tokens):
...
-                                        cell["row_nums"][idx] += 1
+                                cell_row_nums = cell["row_nums"]
+                                if (
+                                    row_idx in cell_row_nums and row_idx + 1 in cell_row_nums
+                                ):  # if the cell spans the 2 rows increase the span
+                                    for idx, row_num in enumerate(range(len(cell_row_nums), 0, -1)):
+                                        cell_row_nums[idx] += 1
+                                        if row_num <= (row_idx + 1):
+                                            cell_row_nums.insert(idx, row_idx + 1)
+                                            break
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1083]: wait why not just `cell_row_nums.append(max(cell_row_nums) + 1)`? Also I think this iterator is crazy:
0, 3
1, 2
2, 1
3, 0

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py [hunk 6]
```diff
@@ -1045,7 +1047,95 @@ def structure_to_cells(table_structure, tokens):
         cell_rect = row_rect.intersect(column_rect)
         if cell_rect.area > 0:
             cell["bbox"] = cell_rect.to_list()
-
+    if union_tokens:
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1050]: Can I get a block of comment describing the algorithm we're implementing?

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py (hunk 5 above)
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py]: Changed, I forgot that all span row numbers must be continuous

File: lib/sycamore/sycamore/transforms/table_structure/table_transformers.py (hunk 6 above)
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1134]: Done.

Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/table_structure/table_transformers.py:1134]: Done.

**Review by HenryL27 — APPROVED** [review]
At least doesn't crash on all the docs in my debugging directory

**Pull request merged**